GMMK Pro Keymap with strong Apple support

seajaysec opened the pull request:
## Description

This keymap for the GMMK Pro focuses on color-coded layers for: Windows/Linux, macOS, and a pseudo-numpad. The inspiration for this keymap, vs using mike1808's (which this is built on) is to provide a first-class macOS experience. That is my daily driver OS and none of the extant keymaps at this time (Late Summer 2021) provide support for Apple device functions such as brightness or media playback control. If you are familiar with using an Apple keyboard, those keys are now all roughly where you'd expect them, with some slight positioning changes to account for the row 1 splits.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* N/A

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

zvecr reviewed (changes requested):
All `.png` files need to be removed as they are not accepted to this repo. Hosting externally on something like imgur is usually the recommend way forward.

drashna commented on keyboards/gmmk/pro/ansi/keymaps/seajaysec/README.md:
```diff
@@ -0,0 +1,68 @@
+QMK layout for gmmk/pro
+=======================
```
```suggestion
# QMK layout for gmmk/pro
```

drashna commented on keyboards/gmmk/pro/ansi/keymaps/seajaysec/config.h:
```diff
@@ -0,0 +1,87 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#define COMBO_COUNT 1
```
```suggestion
#pragma once

#define COMBO_COUNT 1
```

drashna commented on keyboards/gmmk/pro/ansi/keymaps/seajaysec/rules.mk:
```diff
@@ -0,0 +1,26 @@
+VIA_ENABLE = no
+WPM_ENABLE = no
+
+SRC += utils.c
+SRC += seajaysec.c
+SRC += process_record.c
+SRC += encoder.c
```
```suggestion
SRC += utils.c eajaysec.c process_record.c encoder.c
```

seajaysec commented:
> All `.png` files need to be removed as they are not accepted to this repo. Hosting externally on something like imgur is usually the recommend way forward.

Removed the `.png` files and replaced with imgur links. Thanks for the hosting recommendation.

seajaysec reviewed (commented):
Improved mac key layout to right of spacebar.

tzarc commented on keyboards/gmmk/pro/ansi/keymaps/seajaysec/seajaysec.json:
```diff
@@ -0,0 +1,12 @@
+{
```
Please remove this file -- you've got `keymap.c` already.

tzarc commented on keyboards/gmmk/pro/ansi/keymaps/seajaysec/config.h:
```diff
@@ -0,0 +1,89 @@
+    //    #define DISABLE_RGB_MATRIX_GRADIENT_UP_DOWN          // Top down rainbow!
+     #define DISABLE_RGB_MATRIX_GRADIENT_LEFT_RIGHT       // Left Right Rainbow (the top-down is better imho).
+    //    #define DISABLE_RGB_MATRIX_BREATHING                 // Built in (MAPPED!)
+    //    #define DISABLE_RGB_MATRIX_BAND_SAT                  // Single hue band fading saturation scrolling left to right (with white)
+    //    #define DISABLE_RGB_MATRIX_BAND_VAL                  // Single hue band fading brightness scrolling left to right (with black)
+     #define DISABLE_RGB_MATRIX_BAND_PINWHEEL_SAT         // Single hue 3 blade spinning pinwheel fades sat (RGB_MATRIX_CYCLE_PINWHEEL)
+     #define DISABLE_RGB_MATRIX_BAND_PINWHEEL_VAL         // See above
+     #define DISABLE_RGB_MATRIX_BAND_SPIRAL_SAT           // See above
+    //    #define DISABLE_RGB_MATRIX_BAND_SPIRAL_VAL           // Single hue spinning spiral fades brightness (with black)
+     #define DISABLE_RGB_MATRIX_CYCLE_ALL                 // Full keyboard cycling through rainbow
+    //    #define DISABLE_RGB_MATRIX_CYCLE_LEFT_RIGHT          // BUILT IN (MAPPED!) Full Rainbow Left to Right
+     #define DISABLE_RGB_MATRIX_CYCLE_UP_DOWN             // Full gradient scrolling top to bottom
+     #define DISABLE_RGB_MATRIX_CYCLE_OUT_IN              // Rainbow circles coming to center.
+     #define DISABLE_RGB_MATRIX_CYCLE_OUT_IN_DUAL         // Two Rainbow circles coming to 1/3 and 2/3 points.
+     #define DISABLE_RGB_MATRIX_RAINBOW_MOVING_CHEVRON    // Full gradent Chevron shapped scrolling left to right
+    //    #define DISABLE_RGB_MATRIX_CYCLE_PINWHEEL            // BUILT IN (MAPPED!) - Consider disabling this - my poor eyes.
+     #define DISABLE_RGB_MATRIX_CYCLE_SPIRAL              // Full gradient spinning spiral around center of keyboard
+     #define DISABLE_RGB_MATRIX_DUAL_BEACON               // Two rainbows spinning around keyboards
+     #define DISABLE_RGB_MATRIX_RAINBOW_BEACON            // Similar to above.
+     #define DISABLE_RGB_MATRIX_RAINBOW_PINWHEELS         // Enough of the spinning rainbows - we get it! :-)
+    //    #define DISABLE_RGB_MATRIX_RAINDROPS                 // Jellybeans are better.
+     #define DISABLE_RGB_MATRIX_JELLYBEAN_RAINDROPS       // Randomly changes a single key's hue and saturation - better than above.
+     #define DISABLE_RGB_MATRIX_HUE_BREATHING             // Hue shifts up a slight ammount at the same time, then shifts back
+     #define DISABLE_RGB_MATRIX_HUE_PENDULUM              // Hue shifts up a slight ammount in a wave to the right, then back to the left
+     #define DISABLE_RGB_MATRIX_HUE_WAVE                  // Hue shifts up a slight ammount and then back down in a wave to the right
+    //    #define DISABLE_RGB_MATRIX_TYPING_HEATMAP            // Hot keys!
+    //    #define DISABLE_RGB_MATRIX_DIGITAL_RAIN              // The Matrix!
+     #define DISABLE_RGB_MATRIX_SOLID_REACTIVE_SIMPLE     // Key hits shown in current hue - all other keys black: wide slightly better imo
+     #define DISABLE_RGB_MATRIX_SOLID_REACTIVE            // Key hits shown in current hue - all other keys shifted hue
+    //    #define DISABLE_RGB_MATRIX_SOLID_REACTIVE_WIDE       // Hue & value pulse around a single key hit then fades value out (black otherwise)
+     #define DISABLE_RGB_MATRIX_SOLID_REACTIVE_MULTIWIDE  // same as above but a bit too intense imo.
+     #define DISABLE_RGB_MATRIX_SOLID_REACTIVE_CROSS      // Column and Row single color fade.
+     #define DISABLE_RGB_MATRIX_SOLID_REACTIVE_MULTICROSS // Column and Row single color fade. (Multi-keys)
+     #define DISABLE_RGB_MATRIX_SOLID_REACTIVE_NEXUS      // Hue & value pulse away on the same column and row of key hit then fades
+     #define DISABLE_RGB_MATRIX_SOLID_REACTIVE_MULTINEXUS // Hue & value pulse away on the same column and row of multi-key hit then fades
+     #define DISABLE_RGB_MATRIX_SPLASH                    // Full rainbow pulses from key hit. All else black.
+     #define DISABLE_RGB_MATRIX_MULTISPLASH               // Full rainbow pulses from multi-keys. All else black.
+     #define DISABLE_RGB_MATRIX_SOLID_SPLASH              // Single color pulses from key hit. All else black. (Using solid_reactive_wide instead).
+     #define DISABLE_RGB_MATRIX_SOLID_MULTISPLASH         // Single color pulses from muli-keys. All else black. (Using solid_reactive_wide instead).
+#endif
```
These are now `ENABLE_xxxxx` as per the [docs](https://docs.qmk.fm/#/feature_rgb_matrix?id=rgb-matrix-effects).

tzarc commented on keyboards/gmmk/pro/ansi/keymaps/seajaysec/encoder.c:
```diff
@@ -0,0 +1,111 @@
+
+static uint8_t state = ENCODER_DEFAULT;
+
+// clang-format off
+const encoder_callback encoder_mapping[][2] = {
+    [ENCODER_VOLUME] = {&volume_up, &volume_down},
+#ifdef RGB_MATRIX_ENABLE
+    [ENCODER_RGB_HUE] = {&rgb_matrix_increase_hue_noeeprom, &rgb_matrix_decrease_hue_noeeprom},
+    [ENCODER_RGB_SAT] = {&rgb_matrix_increase_sat_noeeprom, &rgb_matrix_decrease_sat_noeeprom},
+    [ENCODER_RGB_VAL] = {&rgb_matrix_increase_val_noeeprom, &rgb_matrix_decrease_val_noeeprom},
+    [ENCODER_RGB_EFFECT] = {&rgb_matrix_step_noeeprom, &rgb_matrix_step_reverse_noeeprom},
+    [ENCODER_RGB_EFFECT_SPEED] = {&rgb_matrix_increase_speed_noeeprom, &rgb_matrix_decrease_speed_noeeprom},
+#endif // RGB_MATRIX_ENABLE
+};
```
Just a heads-up for future -- the `develop` branch (which should be merged in May) now has the capability of specifying an encoder map much like keymap, with normal `KC_xxx` keycodes, pumping through normal `process_record_user()`: [devel docs](https://qmk.github.io/qmk_docs_devel/#/feature_encoders?id=encoder-map).

tzarc commented on keyboards/gmmk/pro/ansi/keymaps/seajaysec/rules.mk:
```diff
@@ -0,0 +1,23 @@
+
+SRC += utils.c seajaysec.c process_record.c encoder.c
+
+ifeq ($(strip $(WPM_ENABLE)), yes)
+	SRC += fun.c
+endif
+
+ifeq ($(strip $(RGB_MATRIX_LEDMAPS)), yes)
+	SRC += rgb_matrix_ledmaps.c
+	OPT_DEFS += -DRGB_MATRIX_LEDMAPS_ENABLED
+endif
```
These should probably be in a `post_rules.mk` instead -- `rules.mk` will soon require assignments only.

andrebrait commented on keyboards/gmmk/pro/ansi/keymaps/seajaysec/encoder.c:
```diff
@@ -0,0 +1,111 @@
+        (*encoder_mapping[state][1])();
+    }
+
+    return true;
```
This will cause the volume keys to be pressed as well as the this handling, btw. Volume key taps have been added to `pro.c` in e335d62e.

noroadsleft reviewed (changes requested):
All of the files in this pull request need to be moved to `keyboards/gmmk/pro/rev1/ansi/keymaps/seajaysec/` – the keyboard's source was moved in the last `develop` branch merge.

drashna commented:
Appears abandoned.  Closing.
